File: engine/src/model-gateway/providers.ts
import { ProviderDescriptor } from '../../../shared';

// LyaCodeX — providers oficiais
// lyacodex-trial removido: trial.lyacodex.ai ainda não existe.
// Adicionar quando o servidor estiver hospedado.

export const LYA_PROVIDERS: ProviderDescriptor[] = [
  {
    id: 'ollama',
    name: 'Ollama (OpenAI GPT Local)',
    kind: 'local',
    defaultBaseUrl: 'http://localhost:11434/v1',
    requiresKeyRef: false,
    supportsStreaming: true,
    openAiCompatible: true,
  },
  {
    id: 'lmstudio',
    name: 'LM Studio',
    kind: 'local',
    defaultBaseUrl: 'http://localhost:1234/v1',
    requiresKeyRef: false,
    supportsStreaming: true,
    openAiCompatible: true,
  },
  {
    id: 'openrouter',
    name: 'OpenRouter',
    kind: 'cloud',
    defaultBaseUrl: 'https://openrouter.ai/api/v1',
    requiresKeyRef: true,
    supportsStreaming: true,
    openAiCompatible: true,
  },
  {
    id: 'openai',
    name: 'OpenAI',
    kind: 'cloud',
    defaultBaseUrl: 'https://api.openai.com/v1',
    requiresKeyRef: true,
    supportsStreaming: true,
    openAiCompatible: true,
  },
  {
    id: 'gemini',
    name: 'Google Gemini',
    kind: 'cloud',
    defaultBaseUrl: 'https://generativelanguage.googleapis.com/v1beta/openai',
    requiresKeyRef: true,
    supportsStreaming: true,
    openAiCompatible: true,
  },
  {
    id: 'anthropic',
    name: 'Anthropic Claude',
    kind: 'cloud',
    defaultBaseUrl: 'https://api.anthropic.com/v1',
    requiresKeyRef: true,
    supportsStreaming: true,
    openAiCompatible: false,
  },
  {
    id: 'groq',
    name: 'Groq',
    kind: 'cloud',
    defaultBaseUrl: 'https://api.groq.com/openai/v1',
    requiresKeyRef: true,
    supportsStreaming: true,
    openAiCompatible: true,
  },
  {
    id: 'custom-openai',
    name: 'Custom OpenAI-Compatible',
    kind: 'custom',
    requiresKeyRef: false,
    supportsStreaming: true,
    openAiCompatible: true,
  },
];

export function getProvider(providerId: string) {
  return LYA_PROVIDERS.find((p) => p.id === providerId);
}

File: backend/src/providers.rs
use crate::contracts::{ProviderDescriptor, ProviderKind};

pub fn list_providers() -> Vec<ProviderDescriptor> {
    vec![
        ProviderDescriptor {
            id: "ollama".into(),
            name: "Ollama (OpenAI GPT Local)".into(),
            kind: ProviderKind::Local,
            default_base_url: Some("http://localhost:11434/v1".into()),
            requires_key_ref: false,
            supports_streaming: true,
            openai_compatible: true,
        },
        ProviderDescriptor {
            id: "lmstudio".into(),
            name: "LM Studio".into(),
            kind: ProviderKind::Local,
            default_base_url: Some("http://localhost:1234/v1".into()),
            requires_key_ref: false,
            supports_streaming: true,
            openai_compatible: true,
        },
        // lyacodex-trial removido: servidor ainda não existe.
        // Adicionar quando trial.lyacodex.ai estiver hospedado.
        ProviderDescriptor {
            id: "openai".into(),
            name: "OpenAI".into(),
            kind: ProviderKind::Cloud,
            default_base_url: Some("https://api.openai.com/v1".into()),
            requires_key_ref: true,
            supports_streaming: true,
            openai_compatible: true,
        },
        ProviderDescriptor {
            id: "openrouter".into(),
            name: "OpenRouter".into(),
            kind: ProviderKind::Cloud,
            default_base_url: Some("https://openrouter.ai/api/v1".into()),
            requires_key_ref: true,
            supports_streaming: true,
            openai_compatible: true,
        },
        ProviderDescriptor {
            id: "gemini".into(),
            name: "Google Gemini".into(),
            kind: ProviderKind::Cloud,
            default_base_url: Some("https://generativelanguage.googleapis.com/v1beta/openai".into()),
            requires_key_ref: true,
            supports_streaming: true,
            openai_compatible: true,
        },
        ProviderDescriptor {
            id: "anthropic".into(),
            name: "Anthropic Claude".into(),
            kind: ProviderKind::Cloud,
            default_base_url: Some("https://api.anthropic.com/v1".into()),
            requires_key_ref: true,
            supports_streaming: true,
            openai_compatible: false,
        },
        ProviderDescriptor {
            id: "groq".into(),
            name: "Groq".into(),
            kind: ProviderKind::Cloud,
            default_base_url: Some("https://api.groq.com/openai/v1".into()),
            requires_key_ref: true,
            supports_streaming: true,
            openai_compatible: true,
        },
        ProviderDescriptor {
            id: "custom-openai".into(),
            name: "Custom OpenAI-Compatible".into(),
            kind: ProviderKind::Custom,
            default_base_url: None,
            requires_key_ref: false,
            supports_streaming: true,
            openai_compatible: true,
        },
    ]
}

pub fn get_provider(provider_id: &str) -> Option<ProviderDescriptor> {
    list_providers()
        .into_iter()
        .find(|provider| provider.id == provider_id)
}
